### ISSUE TITLE
Some of Configuration.browserCapabilities are lost

### ISSUE THREAD

**Issue by VadimGulich** [selenide/selenide#676]
Are not added capabilities chrome
Selenide: 4.10

Examle code:
```
@BeforeSuite
public void setUp() {
  ChromeOptions chromeOptions = new ChromeOptions();
  chromeOptions.addArguments(("load-extension=C:\Users\extension"));
  chromeOptions.addExtensions(new File("/path/to/extension.crx"));
  Configuration.browserCapabilities = new DesiredCapabilities();
  Configuration.browserCapabilities.setCapability(ChromeOptions.CAPABILITY, chromeOptions);
  Configuration.browser = (getPropertyConfig("browser"));
}
```

@BorisOsipov 11:09:
@VadimGulich да проблема есть, сделаешь ишью на гитхабе?

@VadimGulich 11:09: 
@BorisOsipov Да, сделаю

**Comment by gerasimenko-ivan** [comment]
There is a workaround for this issue: you can define WebDriver manually and use com.codeborne.selenide.WebDriverRunner.setWebDriver() method to use WebDriver with ChromeOptions you want, e.g.:

```
        ChromeOptions options = new ChromeOptions();
        options.addArguments("user-data-dir=" + chromeProfile);
        options.addArguments("start-maximized");

        System.setProperty("webdriver.chrome.driver", "C:\\temp\\chromedriver.exe");
        WebDriver webDriver = new ChromeDriver(options);
        webDriver.navigate().to("http://zao-t.wl.test.eirc.mos.ru:7003/pir_war/#");

        setWebDriver(webDriver);
```

BTW: if you want to change user profile (like in the code above) you'd probably need to make a special Automation profile otherwise you can run you tests only when all the instances of browser with profile you use are closed.

**Comment by 13vinni** [comment]
I am trying to update the default browser download settings using the below steps:
               ChromeOptions options = new ChromeOptions();
		HashMap<String, Object> chromePref = new HashMap<>();
		chromePref.put("download.default_directory", <Download folder Location>);
		chromePref.put("profile.default_content_settings.popups", 0);
		options.setExperimentalOption("prefs", chromePref);
		DesiredCapabilities capabilities = new DesiredCapabilities();
		capabilities.setCapability(ChromeOptions.CAPABILITY, options);
		options.merge(capabilities);
		Configuration.browserCapabilities = capabilities;
It does not work for me? 

Is this the same issue as in this thread.

**Comment by dmorokov** [comment]
The same issue found in Selenide 5.6.0

**Comment by ykarpovets** [comment]
According to the java doc in https://github.com/selenide/selenide/blob/master/statics/src/main/java/com/codeborne/selenide/Configuration.java this is possible to achieve via system property "chromeoptions.prefs", so you can 
```
System.setProperty("chromeoptions.prefs", "profile.default_content_settings.popups=0,download.default_directory=<Download folder Location>"));
```

**Comment by Firefds** [comment]
Using Selehide 5.10.0, this also doesn't seem to work at all

```
Configuration.baseUrl = url;
ChromeOptions options = new ChromeOptions();
options.addArguments("--disable-extensions");
options.addArguments("--headless");
options.addArguments("--no-sandbox");
options.addArguments("--window-size=1920,1080");
DesiredCapabilities capabilities = new DesiredCapabilities();
capabilities.setCapability(ChromeOptions.CAPABILITY, options);
Configuration.browserCapabilities = capabilities;
open(url);
```

Even when minimizing down to just
```
Configuration.baseUrl = url;
ChromeOptions options = new ChromeOptions();
options.addArguments("--window-size=1920,1080");
DesiredCapabilities capabilities = new DesiredCapabilities();
capabilities.setCapability(ChromeOptions.CAPABILITY, options);
Configuration.browserCapabilities = capabilities;
open(url);
```

The workaround mentioned in https://github.com/selenide/selenide/issues/676#issuecomment-407321797 works, but I want to avoid calling setWebDriver manually

**Comment by anplay** [comment]
Hi friends,

Any updates about the fix and release that?
It is a bit affect current implementation.

Thank you and have a nice day!

**Comment by asolntsev** [comment]
@anplay Since it's cause by Selenium issue https://github.com/SeleniumHQ/selenium/issues/5279, we don't know when they are going to fix it. Everyone of us can also do it - it's open source. 

@anplay If you have any ideas how to overcome it on Selenide side - you are welcome. Please share them.

**Comment by Firefds** [comment]
@asolntsev 
I went over the Selenium code. The bug seems to be in the `ChromeOptions` class, `asMap` method.
https://github.com/SeleniumHQ/selenium/blob/e82be7d3584062a0d16af8d562d387e3d4855aa1/java/client/src/org/openqa/selenium/chrome/ChromeOptions.java#L268

The "goog:chromeOptions" of the original capabilities are being overwritten by this method in this line -
https://github.com/SeleniumHQ/selenium/blob/e82be7d3584062a0d16af8d562d387e3d4855aa1/java/client/src/org/openqa/selenium/chrome/ChromeOptions.java#L294

As a workaround on Selenide side, we could perhaps add the custom arguments in the `ChromeDriverFactory` class, after the line https://github.com/selenide/selenide/blob/bf2b214ec5f2c002ee403cdffc21d8b3d8b98cc7/src/main/java/com/codeborne/selenide/webdriver/ChromeDriverFactory.java#L43 and before merging the options, as this is the only argument that is not being overwritten by the `asMap` method

**Comment by BorisOsipov** [comment]
@Firefds to be honest didn't get what workaround you suggested. Can I ask you to share some code snippet with changes in `ChromeDriverFactory` you mean?

**Comment by Firefds** [comment]
@BorisOsipov this seems to work. It's raw and can probably be better, but the chrome options I pass finally work correctly - 

```
options.addArguments("--proxy-bypass-list=<-loopback>");
TreeMap<?, ?> capabilities = (TreeMap<?, ?>) config.browserCapabilities().getCapability(ChromeOptions.CAPABILITY);
List<?> list = (List<?>) capabilities.get("args");
for (Object argument : list) {
      options.addArguments(argument.toString());
}
options.merge(createCommonCapabilities(config, proxy));
```

works with
```
Configuration.baseUrl = url;
ChromeOptions options = new ChromeOptions();
options.addArguments("--disable-extensions");
options.addArguments("--headless");
options.addArguments("--no-sandbox");
options.addArguments("--window-size=1920,1080");
DesiredCapabilities capabilities = new DesiredCapabilities();
capabilities.setCapability(ChromeOptions.CAPABILITY, options);
Configuration.browserCapabilities = capabilities;
open("/");
```

**Comment by BorisOsipov** [comment]
@Firefds yes it will work. But only for chrome args. To completely fix issue we will must do the same in all `*BrowserName*DriverFactories` for all `*BrowserName*Options` and their intrenals capabilities - and if we do so we will reimplement `merge` in original selenium `org.openqa.selenium.MutableCapabilities#merge` :) It isn't I want to do(

Does this look correct?

**Comment by BorisOsipov** [comment]
But the more I look at this problem the more I understand taht is only solutions as workaround in selenide.

**Comment by Firefds** [comment]
@BorisOsipov I agree.. the problem is indeed with the `asMap` method of the drivers, for example the ChromeDriver
https://github.com/SeleniumHQ/selenium/blob/e82be7d3584062a0d16af8d562d387e3d4855aa1/java/client/src/org/openqa/selenium/chrome/ChromeOptions.java#L268

But Selenium is working on the new 4 version, and the new Selenium 4 API isn't compatible with Selenide as for now

**Comment by Firefds** [comment]
How about adding a more general field in one of the `Configuration` or `SelenideConfig` classes that stores raw browser arguments?

**Comment by BorisOsipov** [comment]
> new Selenium 4 API isn't compatible with Selenide as for now

which API changes you talk about?

**Comment by Firefds** [comment]
> > new Selenium 4 API isn't compatible with Selenide as for now
> 
> which API changes you talk about?

https://mvnrepository.com/artifact/org.seleniumhq.selenium/selenium-api/4.0.0-alpha-5
and also the new 4 chrome web driver
https://mvnrepository.com/artifact/org.seleniumhq.selenium/selenium-chrome-driver/4.0.0-alpha-5

**Comment by BorisOsipov** [comment]
@Firefds yes we already do some steps to check how to migrate to v4. But now I don't see any changes in this API. 
see this PR 
https://github.com/selenide/selenide/pull/1099

**Comment by BorisOsipov** [comment]
>How about adding a more general field in one of the Configuration or SelenideConfig classes that stores raw browser arguments?

Maybe it is good. How do you see such field should work?

**Comment by Firefds** [comment]
> > How about adding a more general field in one of the Configuration or SelenideConfig classes that stores raw browser arguments?
> 
> Maybe it is good. How do you see such field should work?

Similar to how `ChromeOptions.addArguments` and `Configuration.browserCapabilities` work.

We should have a class called `BrowserArguments` with has a field `List<String> args`.
With `BrowserArguments.addArguments` we could add arguments to that object.
Then pass it to `Configuration` using `Configuration.browserArguments` and add the arguments to the browser options like I mentioned https://github.com/selenide/selenide/issues/676#issuecomment-607403770

**Comment by Firefds** [comment]
@BorisOsipov example here
https://github.com/Firefds/selenide/commit/7aba54bdffcb84b6108217a0f61ad9482d7010cd

**Comment by asolntsev** [comment]
I hope it will be fixed by PR https://github.com/selenide/selenide/pull/1155

**Comment by marisita** [comment]
Hi, is this feature present in Selenide maven version 5.12.2? 
Some classes do not exist (e.g. BrowserArguments) in downloaded Selenide classes after I pull maven dependency, however, the last commit (7aba54bdffcb84b6108217a0f61ad9482d7010cd) in master in Selenide repository contains BrowserArguments.java class.

**Comment by asolntsev** [comment]
Yes, this feature is included in Selenide 5.12.0 and later.

**Comment by vlad8x8** [comment]
looks like broken again (
